### PR TITLE
fix #2268: fix expression parsing bug

### LINKED ISSUE #2268: querying a range that includes the past and the future causes a panic
the following script will cause a panic.  the script is similar to the one added in #2159 with an extra `time > now() - 2d` in the where clause of the last query.

```sh
now=$(date '+%FT%T.000Z')
tomorrow=$(date -v +1d '+%FT%T.000Z')

echo "creating database"
curl -G -ipv4 http://localhost:8086/query --data-urlencode "q=CREATE DATABASE foo"

echo "creating retention policy"
curl -G -ipv4 http://localhost:8086/query --data-urlencode "q=CREATE RETENTION POLICY bar ON foo DURATION INF REPLICATION 1 DEFAULT"

echo "inserting data"
curl -d "{\"database\" : \"foo\", \"retentionPolicy\" : \"bar\", \"points\": [{\"name\": \"cpu\", \"tags\": {\"region\":\"uswest\",\"host\": \"server01\"},\"timestamp\": \"$now\",\"fields\": {\"value\": 100}}]}" -H "Content-Type: application/json" -ipv4 http://localhost:8086/write

echo "inserting data"
curl -d "{\"database\" : \"foo\", \"retentionPolicy\" : \"bar\", \"points\": [{\"name\": \"cpu\", \"tags\": {\"region\":\"uswest\",\"host\": \"server01\"},\"timestamp\": \"$tomorrow\",\"fields\": {\"value\": 200}}]}" -H "Content-Type: application/json" -ipv4 http://localhost:8086/write

sleep 1

echo "querying data"
curl -G -ipv4 http://localhost:8086/query --data-urlencode "db=foo" --data-urlencode "q=SELECT count(value) FROM \"foo\".\"bar\".cpu"

echo "querying data"
curl -G -ipv4 http://localhost:8086/query --data-urlencode "db=foo" --data-urlencode "q=SELECT count(value) FROM \"foo\".\"bar\".cpu where time > now() - 2d and time < now() + 10d"
```

### PR DESCRIPTION
**Pull request by dgnorton**
@benbjohnson 

### Before the fix...

bad

### After the fix...

good

### CODE REVIEW

**Comment by benbjohnson** [comment]
Can you explain the `isroot` variable? I don't understand why it's needed.

**Comment by dgnorton** [comment]
@benbjohnson sometimes it needs to replace the entire expression tree and other times in descents into the tree and modifies a sub node.  I hate this change...I'm sure there's a cleaner way that I'm just not seeing.  This was what I wanted to pair on yesterday.  I have an idea that'll I'll try.  If that doesn't clean it up then maybe we can pair for 15 - 20 min later today.

**Comment by benbjohnson** [comment]
@dgnorton sure, sounds good.

**Comment by toddboom** [comment]
:+1: if you can make ben happy. :dancer:

**Comment by benbjohnson** [comment]
works for me

File: influxql/parser.go [hunk 1]
```diff
@@ -1737,15 +1739,18 @@ func (p *Parser) ParseExpr() (Expr, error) {
...
+		//   precedence >= the operator being added.
+		for node := root; ; {
+			r, ok := node.RHS.(*BinaryExpr)
+			if !ok || r.Op.Precedence() >= op.Precedence() {
+				node.RHS = &BinaryExpr{LHS: node.RHS, RHS: rhs, Op: op}
+				break
 			}
-		} else {
-			expr = &BinaryExpr{LHS: expr, RHS: rhs, Op: op}
+
```
Comment (otoolep) [inline influxql/parser.go]: Nit: looks like a superfluous blank line here.

File: influxql/parser.go [hunk 2]
```diff
@@ -1737,15 +1739,18 @@ func (p *Parser) ParseExpr() (Expr, error) {
...
 		}
 
-		// Assign the new root based on the precendence of the LHS and RHS operators.
-		if lhs, ok := expr.(*BinaryExpr); ok && lhs.Op.Precedence() < op.Precedence() {
-			expr = &BinaryExpr{
-				LHS: lhs.LHS,
-				RHS: &BinaryExpr{LHS: lhs.RHS, RHS: rhs, Op: op},
-				Op:  lhs.Op,
+		// Find the right spot in the tree to add the new expression:
+		//   Descent the RHS of the expression tree until we reach the last
```
Comment (otoolep) [inline influxql/parser.go]: Nit: whitespacing off here?

File: influxql/parser.go [hunk 3]
```diff
@@ -1737,15 +1739,18 @@ func (p *Parser) ParseExpr() (Expr, error) {
...
+		//   BinaryExpr OR a BinaryExpr who's RHS has an operator with
```
Comment (otoolep) [inline influxql/parser.go]: Super-nit: who's -> whose

File: influxql/parser_test.go [hunk 4]
```diff
@@ -1172,6 +1172,32 @@ func TestParser_ParseExpr(t *testing.T) {
 			},
 		},
 
+		// Complex binary expression.
```
Comment (otoolep) [inline influxql/parser_test.go]: +1 on test, looks good to me.

**Comment by otoolep** [comment]
Some super-picky nits, test case looks good. +1 on green build.

**Pull request merged**